### PR TITLE
#3199 Add support for implicit conversion between java.time.LocalDate and java.time.LocalDateTime

### LINKED ISSUE #3199: Support for implicit conversion between java.time.LocalDate and java.time.LocalDateTime
### Expected behavior

Either `LocalDateTime map(LocalDate date)` results in compilation error or mapping results in `date == null ? null : LocalDateTime.of(date, LocalTime.MIN)`

### Actual behavior

`LocalDateTime map(LocalDate date)` maps any `date` to `null`

The problem seems to be related to incorrect mapping sequence: `LocalDate` to `XMLGregorianCalendar` and `XMLGregorianCalendar` to `LocalDateTime`. Separately those mappings are correct, but combined are not:
Generated mapper contains the following line:
```java
LocalDateTime localDateTime = xmlGregorianCalendarToLocalDateTime( localDateToXmlGregorianCalendar( date ) ); 
```

Generated method `xmlGregorianCalendarToLocalDateTime(..)` returns `null` if hours, minutes or seconds are equal to `DatatypeConstants.FIELD_UNDEFINED`, which they are cause `localDateToXmlGregorianCalendar(..)` calls `javax.xml.datatype.DatatypeFactory::newXMLGregorianCalendarDate` which sets hours, minuts, seconds and milliseconds as `FIELD_UNDEFINED`

Methods I'm talking about for reference:
```java
        private static LocalDateTime xmlGregorianCalendarToLocalDateTime( XMLGregorianCalendar xcal ) {
            if ( xcal == null ) {
                return null;
            }

            if ( xcal.getYear() != DatatypeConstants.FIELD_UNDEFINED
                && xcal.getMonth() != DatatypeConstants.FIELD_UNDEFINED
                && xcal.getDay() != DatatypeConstants.FIELD_UNDEFINED
                && xcal.getHour() != DatatypeConstants.FIELD_UNDEFINED
                && xcal.getMinute() != DatatypeConstants.FIELD_UNDEFINED
            ) { //some mapping }
            return null;
        }
```

```java
private XMLGregorianCalendar localDateToXmlGregorianCalendar( LocalDate localDate ) {
            if ( localDate == null ) {
                return null;
            }

            return datatypeFactory.newXMLGregorianCalendarDate(
                localDate.getYear(),
                localDate.getMonthValue(),
                localDate.getDayOfMonth(),
                DatatypeConstants.FIELD_UNDEFINED );
        }
```

```java
    public XMLGregorianCalendar newXMLGregorianCalendarDate(
            final int year,
            final int month,
            final int day,
            final int timezone) {

            return newXMLGregorianCalendar(
                    year,
                    month,
                    day,
                    DatatypeConstants.FIELD_UNDEFINED, // hour
                    DatatypeConstants.FIELD_UNDEFINED, // minute
                    DatatypeConstants.FIELD_UNDEFINED, // second
                    DatatypeConstants.FIELD_UNDEFINED, // millisecond
                    timezone);
            }
```

### Steps to reproduce the problem

Define Mapper:
```java
@Mapper
public interface IssueBelkinmikeMapper {

    IssueBelkinmikeMapper INSTANCE = Mappers.getMapper( IssueBelkinmikeMapper.class );

    LocalDateTime map(LocalDate date);
}
```

Define Test:
```java
@WithClasses({
    IssueBelkinmikeMapper.class
})
@IssueKey("belkinmike")
class IssueBelkinmikeMapperTest {

    @ProcessorTest
    void throwsException() {
        LocalDate source = LocalDate.of(2022, 2, 2);
        LocalDateTime target = IssueBelkinmikeMapper.INSTANCE.map(source);
        assertThat(target).isNotNull();
    }
}
```

Test should pass or fail at mapper generation stage, but it is not

IMHO: It should fail at generating such mapping, cause I met this problem mapping dtos:

I had this classes
```java
class EntityDto {
    LocalDate dateFrom;
}
class Entity {
    LocalDate dateFrom;
}

interface MyMapper {
    EntityDto map(Entity e);
}
```

then dto field's type changed to LocalDateTime: 

```java
class EntityDto {
    LocalDateTime dateFrom;
}
class Entity {
    LocalDate dateFrom;
}

interface MyMapper {
    EntityDto map(Entity e);
}
```

And I would expect mapping to fail and attract my attention to it, but it did not

### MapStruct Version

main branch (commit id=a7ba12676d237957be0753076f436d7974c9f7f4)

### PR DESCRIPTION
**Pull request by Bragolgirith**
Fixes #3199 

Took the liberty of creating a PR.
First time contributing to the project, so do tell if I've missed something!

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java [hunk 1]
```diff
@@ -233,12 +233,15 @@ private void registerJava8TimeConversions() {
...
         register( Duration.class, String.class, new StaticParseToStringConversion() );
 
-        // Java 8 to Date
+        // Java 8 time to Date
         register( ZonedDateTime.class, Date.class, new JavaZonedDateTimeToDateConversion() );
         register( LocalDateTime.class, Date.class, new JavaLocalDateTimeToDateConversion() );
         register( LocalDate.class, Date.class, new JavaLocalDateToDateConversion() );
         register( Instant.class, Date.class, new JavaInstantToDateConversion() );
 
+        // Java 8 time (internal)
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java:242]: What do you mean with "(internal)"?

**Review by filiphr — APPROVED** [review]
Thanks @Bragolgirith. It looks good to me. I only have one small question

File: processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java (hunk 1 above)
Comment (Bragolgirith) [reply processor/src/main/java/org/mapstruct/ap/internal/conversion/Conversions.java]: Oh, what I meant was we are converting from a Java 8 type to another Java 8 type, i.e. "internally" within the same package. I've cleaned that up now :)

**Comment by Bragolgirith** [comment]
Also, there were some CI workflow run [failures](https://github.com/mapstruct/mapstruct/actions/runs/4800410646) last time, e.g.:
```
Could not transfer artifact ... (truncated) ... Connect to repo.maven.apache.org:443 [repo.maven.apache.org/146.75.28.215] failed: Connection timed out -> [Help 1]
```
These seem unrelated to the PR - maybe a one-time connectivity issue or a proxy configuration issue?

**Comment by filiphr** [comment]
Thanks for your changes @Bragolgirith.

>Also, there were some CI workflow run [failures](https://github.com/mapstruct/mapstruct/actions/runs/4800410646) last time, e.g.:

Those are some recurring flaky errors we are getting on our CI. I ignore those errors. I'll merge this once the final build is done

**Pull request merged**